elastic/kibana
# [data views] composite runtime fields in api layer

## Summary

Add composite runtime fields to the data view api layer. Broken out of https://github.com/elastic/kibana/pull/110223

Part of phase 3 of https://github.com/elastic/kibana-team/issues/307 

Composite runtime field ES docs - https://www.elastic.co/guide/en/elasticsearch/reference/current/runtime-examples.html#runtime-examples-grok-composite

---

To test composite runtime field, provide data view id  - 

```curl -u elastic -X POST "localhost:5601/api/data_views/data_view/${id}/runtime_field" -H 'kbn-xsrf: true' -H 'Content-Type: application/json' -d @"composite_runtime.txt"```

composite_runtime.txt content - 
```
{
  "name": "composite_runtime",
  "runtimeField": {
    "type": "composite",
    "script": {
      "source": "emit(\"a\",\"a\"); emit(\"b\",\"b\")"
    },
    "fields": {
      "a": {
        "type": "keyword"
      },
      "b": {
        "type": "keyword"
      }
    }
  }
}

```

and delete 
```
curl -u elastic -X DELETE "localhost:5601/api/data_views/data_view/${id}/runtime_field/composite_runtime" -H 'kbn-xsrf: true' -H 'Content-Type: application/json'
```

---

Testing
- Use API to submit composite runtime field
- Verify fields are in data view management. 
- Verify they cannot be deleted
- Verify they can be edited like mapped fields - can't change name or type, no set script ability
- Verify regular runtime fields work as expected
- Go to discover and verify that runtime fields work for search and query as expected
- Remove composite runtime field via API
- verify that all child fields have been removed.
